File: 0015-3sum/0015-3sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        #optimal
        n=len(nums)
        nums.sort()
        ans=[]
        for i in range(n):
            if nums[i]>0:
                break
            elif i!=0 and nums[i]==nums[i-1]:
                    continue
            j=i+1
            k=n-1
            while j<k:              
                s=nums[i]+nums[j]+nums[k]
                if s>0:                   
                    k-=1
                elif s<0:
                    j+=1
                else:
                    ans.append([nums[i],nums[j],nums[k]])
                    j+=1
                    while j<k and nums[j]==nums[j-1]:
                        j+=1
                    k-=1
                    while j<k and nums[k]==nums[k+1]:
                        k-=1
                    
        return ans
#         #better but time limit exceed
#         h=set()
#         for i in range(len(nums)):
#             s=set()
#             for j in range(i+1,len(nums)):
#                 comp=-(nums[i]+nums[j])
#                 if comp in s:
#                     temp = [nums[i],nums[j],comp]
#                     temp.sort()
#                     h.add(tuple(temp))
#                 s.add(nums[j])
        
#         return list(h)
                

File: 0015-3sum/0015-3sum.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        List <List<Integer>> ans = new ArrayList<>();
        Arrays.sort(nums);
        int n = nums.length;
        for(int i=0;i<n;i++){
            if(nums[i]>0)break;
            else if(i!=0 && nums[i]==nums[i-1])continue;
            
            
            int j=i+1,k=n-1;
            
            while(j<k){
                int sum = nums[i]+nums[j]+nums[k];
                if(sum<0)j++;
                
                else if(sum>0)k--;   
                
                else{
                    List<Integer> temp = Arrays.asList(nums[i],nums[j], nums[k]);
                    ans.add(temp);
                    j++;
                    while(j<k && nums[j]==nums[j-1])j++;
                    k--;
                    while(j<k && nums[k]==nums[k+1])k--;
                }
            
        }
    }return ans;
}}